### PR TITLE
New Calibration Code MS^2

### PR DESCRIPTION
**Pull request by MagdaSkurzok**

### CODE REVIEW

File: CalibrationMS/TimeCalibration.cpp [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+	      int thr = thr_time_pair.first;
+	      RefTimeLead[thr] = thr_time_pair.second;
+
+	    }
+	    for(auto & thr_time_pair : trail_times_B){
+	      int thr = thr_time_pair.first;
+	      RefTimeTrail[thr] = thr_time_pair.second;
+	      
+	    }
+	    fRefTimesL.push_back(RefTimeLead[1]);
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: Don't use magic numbers, also it seems you are using only first threshold time, then it is no needed to save all thresholds, or it is prepared for later?

File: CalibrationMS/TimeCalibration.cpp [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+		fRefTimesT.clear();
+	        fhitsCalib.push_back(*hit);
+//
+//If the first hit in the next time window is the refference detector hit save its times
+		if(PMid==kPMidRef){
+		  auto lead_times_B = hit->getSignalB().getRecoSignal().getRawSignal().getTimesVsThresholdNumber(JPetSigCh::Leading);
+		  auto trail_times_B = hit->getSignalB().getRecoSignal().getRawSignal().getTimesVsThresholdNumber(JPetSigCh::Trailing);
+		  for(auto & thr_time_pair : lead_times_B){
+		    int thr = thr_time_pair.first;
+		    RefTimeLead[thr] = thr_time_pair.second;
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: same as above

File: CalibrationMS/TimeCalibration.cpp [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+		  }
+		  for(auto & thr_time_pair : trail_times_B){
+		    int thr = thr_time_pair.first;
+		    RefTimeTrail[thr] = thr_time_pair.second;
+		  }
+		  fRefTimesL.push_back(RefTimeLead[1]);
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: above

File: CalibrationMS/TimeCalibration.cpp [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+//getting Layer and Slots numbers from analysing files name
+std::vector<int> vectorOfNumbers;
+std::string str(gDirectory->GetFile()->GetName());//getting directory of analysing file
+
+int number=0;
+double frac_err=0.3;  //maximal fractional uncertainty accepted by calibration
+int min_ev = 10;     //minimal number of events for a distribution to be fitted
+ for (unsigned int i=str.size()-1;i>0; i--){
+
+	if (isdigit(str[i])){
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: I think, better solution here is std::stoi: http://en.cppreference.com/w/cpp/string/basic_string/stol

File: CalibrationMS/TimeCalibration.cpp [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+}
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+// save timeDiffAB mean values for each slot and each threshold in a JPetAuxilliaryData object
+	// so that they are available to the consecutive modules
+	getAuxilliaryData().createMap("timeDiffAB mean values");
+//
+//create output txt file with calibration parameters 
+//
+	std::ofstream results_fit;
+        results_fit.open(OutputFile, std::ios::app); //file will be overwritten
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: comment says that file will be overwritten, but std::ios::app seek to the end of stream before each write and all output is added at end of file

File: CalibrationMS/TimeCalibration.cpp [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+			double mean_Ref_t = getStatistics().getHisto1D(histo_name_Ref_t).GetMean();
+			getAuxilliaryData().setValue("timeDiffref mean values", histo_name_Ref_t, mean_Ref_t);
+
+			TH1F* histoToSave_Ref_trailing = &(getStatistics().getHisto1D(histo_name_Ref_t));
+//non zero histos 
+// slot.first - ID
+// slot.second - wskaznik na JPetBarrelSlot
+//save fit parameters only for layerX and SlotY (taken from analysing files name)
+//fit just for proper slot
+		        int warstwa = (scin.second)->getBarrelSlot().getLayer().getID();
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: change variable name to english

File: CalibrationMS/TimeCalibration.cpp [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+			  if(histoToSave_trailing->GetEntries()<=min_ev){
+			    results_fit << "#WARNING: Statistics used to determine the trailing edge A-B calibration constant was less than "<<min_ev<<" events!"<<endl;
+			  }
+//fit scintilators
+			cout <<"#############" << endl;
+			cout <<"CALIB_INFO: Fitting histogams for layer= " << warstwa << ", slot= "<< scyntylator <<", threshold= "<<thr<<endl; 
+			cout <<"#############" << endl;
+
+			int highestBin_l = histoToSave_leading->GetBinCenter(histoToSave_leading->GetMaximumBin());
+			histoToSave_leading->Fit("gaus","","", highestBin_l-5, highestBin_l+5);
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: Extract fit range to variable

File: CalibrationMS/TimeCalibration.cpp [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+			histoToSave_leading->Draw();
+		
+			int highestBin_t = histoToSave_trailing->GetBinCenter(histoToSave_trailing->GetMaximumBin());
+			histoToSave_trailing->Fit("gaus","","", highestBin_t-5, highestBin_t+5);
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: same as above

File: CalibrationMS/TimeCalibration.cpp [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+			}
+
+			if((position_peak_error_t/position_peak_t)>=frac_err) {
+			results_fit <<"#WFIT: Large uncertainty on the calibration constant!"<<endl;
+			}
+
+//fit reference detector
+
+			int highestBin_Ref_l = histoToSave_Ref_leading->GetBinCenter(histoToSave_Ref_leading->GetMaximumBin());
+			histoToSave_Ref_leading->Fit("gaus","","", highestBin_Ref_l-5, highestBin_Ref_l+5); //range for fit
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: above

File: CalibrationMS/TimeCalibration.cpp [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+			TF1 *fit_Ref_l = histoToSave_Ref_leading->GetFunction("gaus");
+			fit_Ref_l->SetRange(highestBin_Ref_l-20, highestBin_Ref_l+20); //range to draw gaus function
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: above

File: CalibrationMS/TimeCalibration.cpp [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+//leading edge
+
+	for(auto & thr_time_pair : lead_times_A){
+
+		int thr = thr_time_pair.first;
+
+		if( lead_times_B.count(thr) > 0 ){ // if there was leading time at the same threshold at opposite side
+
+			double timeDiffAB_l = lead_times_A[thr] - lead_times_B[thr];
+			timeDiffAB_l/= 1000.; // we want the plots in ns instead of ps
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: multiplication is usually faster then division, so generally it is better to * 0.001 then / 1000.

File: CalibrationMS/TimeCalibration.cpp [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+			int layer_number = fBarrelMap.getLayerNumber( hit.getBarrelSlot().getLayer() );
+			int slot_number = fBarrelMap.getSlotNumber( hit.getBarrelSlot() );
+			getStatistics().getHisto2D(Form("TimeDiffVsID_layer%d_thr%d_leading", layer_number, thr)).Fill( slot_number,
+															timeDiffAB_l);
+//
+//take minimum time difference between Ref and Scint
+			double timeDiffLmin=10000000000000.;
+			for(unsigned int i = 0; i < fRefTimesL.size();i++) { 
+			double timeDiffHit_L = (lead_times_A[thr] + lead_times_B[thr])/2. -fRefTimesL[i];
+			 timeDiffHit_L = timeDiffHit_L/1000.;//ps -> ns
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: same as above

File: CalibrationMS/TimeCalibration.cpp [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+
+
+//trailing
+	for(auto & thr_time_pair : trail_times_A){
+		int thr = thr_time_pair.first;
+
+		if( trail_times_B.count(thr) > 0 ){ // if there was trailing time at the same threshold at opposite side
+
+			double timeDiffAB_t = trail_times_A[thr] - trail_times_B[thr];
+			timeDiffAB_t/= 1000.; // we want the plots in ns instead of ps
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: above

File: CalibrationMS/TimeCalibration.cpp [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+			int layer_number = fBarrelMap.getLayerNumber( hit.getBarrelSlot().getLayer() );
+			int slot_number = fBarrelMap.getSlotNumber( hit.getBarrelSlot() );
+			getStatistics().getHisto2D(Form("TimeDiffVsID_layer%d_thr%d_trailing", layer_number, thr)).Fill( slot_number, timeDiffAB_t);
+															
+// 
+//taken minimal time difference between Ref and Scint
+			double timeDiffTmin=10000000000000.;
+			for(unsigned int i = 0; i < fRefTimesT.size();i++) { 
+                            double timeDiffHit_T = (trail_times_A[thr] + trail_times_B[thr])/2. -fRefTimesT[i];
+			    timeDiffHit_T = timeDiffHit_T/1000.; //ps->ns
```
Comment (grey277) [inline CalibrationMS/TimeCalibration.cpp]: above

File: CalibrationMS/TimeCalibration.cpp [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
+using namespace std;
+
+TimeCalibration::TimeCalibration(const char * name, const char * description):JPetTask(name, description){}
+
+void TimeCalibration::init(const JPetTaskInterface::Options& opts){
+  time_t czas;
```
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: done

File: CalibrationMS/TimeCalibration.cpp [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+	  CalibRun= atof(opts.at(fCalibRunKey).c_str());
+        }
+
+// create histograms for time differences at each slot and each threshold
+
+//	auto scints = getParamBank().getScintillators();
+	time(&czas);
+	//
+	std::ofstream output;
+	output.open(OutputFile,std::ios::app); //file will be overwritten
```
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: ok

File: CalibrationMS/TimeCalibration.cpp [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+	  //taking refference detector hits times (scin=193)
+	  //
+	  int PMid = hit->getSignalB().getRecoSignal().getRawSignal().getPM().getID();
+	
+	  if(PMid==kPMidRef){
+	    auto lead_times_B = hit->getSignalB().getRecoSignal().getRawSignal().getTimesVsThresholdNumber(JPetSigCh::Leading);
+	    auto trail_times_B = hit->getSignalB().getRecoSignal().getRawSignal().getTimesVsThresholdNumber(JPetSigCh::Trailing);                                                   
+	    for(auto & thr_time_pair : lead_times_B){
+	      int thr = thr_time_pair.first;
+	      RefTimeLead[thr] = thr_time_pair.second;
```
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: Is it necessary to change?

File: CalibrationMS/TimeCalibration.cpp (hunk 1 above)
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: We calculate calibration constants for all thresholds (prepared for later).

File: CalibrationMS/TimeCalibration.cpp (hunk 4 above)
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: ok

File: CalibrationMS/TimeCalibration.cpp (hunk 5 above)
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: ok

File: CalibrationMS/TimeCalibration.cpp (hunk 6 above)
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: ok

File: CalibrationMS/TimeCalibration.cpp (hunk 7 above)
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: The range of histograms for different slots is different, therefore we do it in this way.

File: CalibrationMS/TimeCalibration.cpp (hunk 11 above)
Comment (MagdaSkurzok) [inline CalibrationMS/TimeCalibration.cpp]: ok

**Comment by wkrzemien** [comment]
I'm closing this PR, since a new version  is opened, see:
https://github.com/alekgajos/j-pet-framework-examples/pull/1

Still it needs a code review

**Pull request closed without merging**